File: evolution/heal.py
from __future__ import annotations

import os
import shutil
import time
from typing import Any, Callable


class Heal:
    def __init__(
        self,
        llm_complete_fn: Callable[[str], str] | None = None,
        verify_fn: Callable[[str, str], bool] | None = None,
    ) -> None:
        self._llm_complete = llm_complete_fn
        self._verify_fn = verify_fn
        self._heal_history: dict[str, int] = {}

    def diagnose_and_repair(self, target_file: str, error: str) -> bool:
        if not os.path.isfile(target_file):
            return False
        try:
            with open(target_file, encoding="utf-8") as f:
                original_code = f.read()
        except OSError:
            return False

        if self._llm_complete is None:
            return False

        prompt = (
            f"System Failure detected in {target_file}. Error: {error}\n\n"
            f"Original Code:\n{original_code}\n\n"
            f"Task: REPAIR the code to fix this specific bug. Return ONLY the corrected code."
        )
        repaired_code = self._llm_complete(prompt)
        if not repaired_code:
            return False

        if self._verify_fn is not None and not self._verify_fn(target_file, repaired_code):
            return False

        backup_path = target_file + ".bak"
        try:
            shutil.copy2(target_file, backup_path)
            with open(target_file, "w", encoding="utf-8") as f:
                f.write(repaired_code)
        except OSError:
            return False

        self._heal_history[target_file] = self._heal_history.get(target_file, 0) + 1
        return True

    def get_stats(self) -> list[tuple[str, int]]:
        return list(self._heal_history.items())

    def process(self, input_data: Any) -> dict[str, Any]:
        if isinstance(input_data, dict):
            target = input_data.get("target_file", "")
            error = input_data.get("error", "")
            if target and error:
                success = self.diagnose_and_repair(target, error)
                return {"repaired": success, "target": target}
        return {"history": dict(self._heal_history)}

File: evolution_hydra/Heal.ts
import logger from '../utils/logger.js';
import { errorMessage } from '../types/index.js';
import type { QueryEngine } from '../core/QueryEngine.js';
import type { Aegis } from './Aegis.js';
import * as fs from 'fs';
import * as path from 'path';

/**
 * HEAL is the Recursive Self-Healing system of HYDRA.
 * It detects runtime failures in dynamic tools and 
 * automatically attempts to rewrite and refactor the faulty logic.
 */
export class Heal {
  private engine: QueryEngine;
  private aegis: Aegis;
  private healHistory: Map<string, number> = new Map();

  constructor(engine: QueryEngine, aegis: Aegis) {
    this.engine = engine;
    this.aegis = aegis;
    logger.info('[Heal] 🩹 Recursive Self-Repair system activated.');
  }

  /**
   * Called when a tool or system component crashes. 
   * Triggers a "Diagnostics & Mutation" cycle.
   */
  async diagnoseAndRepair(targetFile: string, error: string): Promise<boolean> {
    const filePath = path.resolve(process.cwd(), targetFile);
    if (!fs.existsSync(filePath)) return false;

    logger.error(`[Heal] 🛑 Trauma detected: ${targetFile} (Error: ${error.substring(0, 50)}...)`);
    const originalCode = fs.readFileSync(filePath, 'utf-8');

    try {
      // 1. Diagnostics: The Council analyzes the "injury"
      const repairPrompt = `System Failure detected in ${targetFile}. 
                            Error Context: ${error}
                            Original Code:\n\n${originalCode}\n\n
                            Task: REPAIR the code to fix this specific bug. 
                            Return ONLY the corrected code block. 
                            Use AEGIS-compliant safe code.`;

      let repairedCode = '';
      for await (const chunk of this.engine.query(repairPrompt)) {
        if (chunk.type === 'text') repairedCode += chunk.content;
      }

      // 2. Proof: Verify with AEGIS before applying the "Surgery"
      const proof = await this.aegis.verifyCode(targetFile, repairedCode);
      if (!proof.isSafe) {
        logger.error(`[Heal] 🛡️ AEGIS BLOCKED SURGERY: The proposed fix is unsafe.`);
        return false;
      }

      // 3. Application: Patch the file
      const backupPath = `${filePath}.bak`;
      fs.writeFileSync(backupPath, originalCode);
      logger.info(`[Heal] 💾 Backup created at ${path.basename(backupPath)}`);

      fs.writeFileSync(filePath, repairedCode);
      logger.info(`[Heal] ✨ SURGERY SUCCESSFUL. ${targetFile} has been healed and mutated.`);

      const count = this.healHistory.get(targetFile) || 0;
      this.healHistory.set(targetFile, count + 1);

      return true;

    } catch (err) {
      logger.error(`[Heal] 🚫 Fatal complication during repair: ${errorMessage(err)}`);
      return false;
    }
  }

  getStats() {
    return Array.from(this.healHistory.entries());
  }
}
